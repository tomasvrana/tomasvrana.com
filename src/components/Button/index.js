import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import Link from '../Link'

const Container = styled.span`
  > a {
    background-color: ${({ theme }) => theme.colors.interactive};
    color: ${({ theme }) => theme.colors.buttonColor};
    border: 0.1rem solid ${({ theme }) => theme.colors.interactive};
    border-radius: 0.2rem;
    letter-spacing: 0.1em;
    padding: 0.6rem 1.8rem;
    font-weight: 600;
    transition: background-color 0.2s ease-in-out;
    cursor: pointer;
    display: inline-block;

    &:hover,
    &:focus {
      color: ${({ theme }) => theme.colors.buttonColor};
      background-color: ${({ theme }) => theme.colors.interactiveHighlight};
    }

    &.secondary {
      background-color: ${({ theme }) => theme.colors.secondaryButtonBackground};
      color: ${({ theme }) => theme.colors.secondaryButtonColor};
      border-color: ${({ theme }) => theme.colors.secondaryButtonBackground};

      &:hover,
      &:focus {
        background-color: ${({ theme }) => theme.colors.secondaryButtonBackgroundHighlight};
        color: ${({ theme }) => theme.colors.secondaryButtonColor};
      }
    }

    &.outline {
      border-color: ${({ theme }) => theme.colors.text};
      background-color: transparent;
      color: ${({ theme }) => theme.colors.text};
    }
  }
`

const Button = (props) => {
  if (!props.href && !props.onClick) throw new Error('Button must either have onClick or href prop')
  return (
    <Container className='text-transform-uppercase'>
      <Link
        target={props.target}
        className={[
          props.secondary ? 'secondary' : '',
          props.outline === 'true' ? 'outline' : ''
        ].join(' ')}
        href={props.href || '#'}
        {...props}
      >
        {props.children}
      </Link>
    </Container>
  )
}

Button.propTypes = {
  children: PropTypes.oneOfType([ PropTypes.string, PropTypes.node ]).isRequired,
  onClick: PropTypes.func,
  href: PropTypes.string,
  secondary: PropTypes.bool,
  outline: PropTypes.string,
  target: PropTypes.string
}

Button.defaultProps = {
  secondary: false,
  outline: 'false'
}

export default Button
