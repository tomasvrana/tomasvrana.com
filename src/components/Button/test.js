import React from 'react'
import Button from './'
import { shallow } from 'enzyme'

function getWrapper (props) {
  return shallow(
    <Button {...props}>
      I'm a button
    </Button>
  )
}

describe('<Button />', () => {
  describe('when no href or onClick is defined', () => {
    test('it throws an error', () => {
      expect(() => {
        getWrapper()
      }).toThrowError('Button must either have onClick or href prop')
    })
  })

  test('it renders a primary button correctly', () => {
    expect(getWrapper({
      onClick: jest.fn(),
      secondary: false,
      target: '_blank',
      title: 'Primary'
    })).toMatchSnapshot()
  })

  test('it renders a secondary button correctly', () => {
    expect(getWrapper({
      onClick: jest.fn(),
      secondary: true,
      target: '_blank',
      title: 'Secondary'
    })).toMatchSnapshot()
  })
})
