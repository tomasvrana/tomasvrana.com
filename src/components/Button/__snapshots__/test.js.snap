// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> it renders a primary button correctly 1`] = `
<styled.span
  className="text-transform-uppercase"
>
  <Link
    className=""
    href="#"
    onClick={[MockFunction]}
    secondary={false}
    target="_blank"
    title="Primary"
    tracking={Object {}}
    useLang={true}
  >
    I'm a button
  </Link>
</styled.span>
`;

exports[`<Button /> it renders a secondary button correctly 1`] = `
<styled.span
  className="text-transform-uppercase"
>
  <Link
    className="secondary"
    href="#"
    onClick={[MockFunction]}
    secondary={true}
    target="_blank"
    title="Secondary"
    tracking={Object {}}
    useLang={true}
  >
    I'm a button
  </Link>
</styled.span>
`;
